gno.land/pkg/gnoland/app.go

```diff
@@ -20,34 +22,52 @@ import (
 	"github.com/gnolang/gno/tm2/pkg/store/iavl"
 )
 
-// NewApp creates the GnoLand application.
-func NewApp(rootDir string, skipFailingGenesisTxs bool, logger log.Logger, maxCycles int64) (abci.Application, error) {
-	// Get main DB.
-	db, err := dbm.NewDB("gnolang", dbm.GoLevelDBBackend, filepath.Join(rootDir, "data"))
-	if err != nil {
-		return nil, fmt.Errorf("error initializing database %q using path %q: %w", dbm.GoLevelDBBackend, rootDir, err)
+type CustomAppConfig struct {
+	DB                    dbm.DB
+	GnoRootDir            string
+	SkipFailingGenesisTxs bool
+	Logger                log.Logger
+	MaxCycles             int64
+}
+
+func (c *CustomAppConfig) applyDefault() {
+	if c.Logger == nil {
+		c.Logger = log.NewNopLogger()
+	}
+
+	if c.DB == nil {
+		c.DB = dbm.NewMemDB()
 	}
 
+	if c.GnoRootDir == "" {
+		c.GnoRootDir = guessGnoRootDir()
+	}
+}
+
+// NewApp creates the GnoLand application.
+func NewCustomApp(cfg CustomAppConfig) (abci.Application, error) {
+	cfg.applyDefault()
```

Comment:
instead of applying default, I suggest just validating. this is the kind of constructor that should be fully reviewed by people using it.

Reply:
Or maybe export this ApplyDefault method and do not call it automatically.

Reply:
added both